test/pipelinerun_test.go

```diff
@@ -33,12 +34,102 @@ import (
 	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
 )
 
+// dumpAndTearDown will, when a test fails, print out all CRD objects as YAML
+// in the relevant namespace before tearing the namespace down. It will also
+// get the logs of all pods in the namespace. If the tests do not fail then the
+// namespace is simply torn down.
+func dumpAndTearDown(t *testing.T, logger *logging.BaseLogger, cs *clients, namespace string) {
+	defer tearDown(logger, cs.KubeClient, namespace)
+	if !t.Failed() {
+		return
+	}
+
+	var output []byte
+
+	printOrAdd := func(kind, name string, i interface{}) {
+		bs, err := yaml.Marshal(i)
+		if err != nil {
+			logger.Fatalf("couldn't marshal %s to yaml: %v", kind, err)
+		} else {
+			output = append(output, []byte("\n---\n")...)
+			output = append(output, bs...)
+		}
+	}
+	pos, err := cs.KubeClient.Kube.CoreV1().Pods(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range pos.Items {
+		output = append(output, fmt.Sprintf("Logs for pod %s\n", i.Name)...)
+		ls, err := getAllLogsFromPod(cs.KubeClient.Kube, i.Name, namespace)
+		if err == nil {
+			output = append(output, ls...)
+		}
+	}
+
+	ps, err := cs.PipelineClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range ps.Items {
+		printOrAdd("Pipeline", i.Name, i)
+	}
+
+	pps, err := cs.PipelineParamsClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range pps.Items {
+		printOrAdd("PipelineParamses", i.Name, i)
+	}
+
+	bds, err := cs.BuildClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range bds.Items {
+		printOrAdd("Build", i.Name, i)
+	}
+
+	prs, err := cs.PipelineResourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range prs.Items {
+		printOrAdd("PipelineResource", i.Name, i)
+	}
+
+	prrs, err := cs.PipelineRunClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range prrs.Items {
+		printOrAdd("PipelineRun", i.Name, i)
+	}
+
+	ts, err := cs.TaskClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range ts.Items {
+		printOrAdd("Task", i.Name, i)
+	}
+	trs, err := cs.TaskRunClient.List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+	for _, i := range trs.Items {
+		printOrAdd("TaskRun", i.Name, i)
+	}
+	logger.Info(string(output))
```

Comment:
what do you think about making this a bit more compact with something like a for loop that iterates over a bunch of `list` functions?

Reply:
Unfortunately all of the `List` functions are different types, and there is no common interface for `List`ing between the various clients, so I wasn't able to find a way to shorten this

Reply:
kk, thanks for looking!